### PR TITLE
Add IBM scrapers

### PR DESCRIPTION
**Pull request by hiltol**

### CODE REVIEW

File: src/scrapers/ibmCatalog.ts [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+type CatalogJson = {
+  [name: string]: (CatalogJson | PricingMetaData)[]
+}
+
+type GCPrice = {
+  price: number;
+  quantity_tier: number;
+}
+
+type ResultsWithNext<Results extends Record<string, unknown> = Record<string, unknown>> = {
```
Comment (balazs-marjan) [inline src/scrapers/ibmCatalog.ts:22]: I think these opaque records make the code very hard to follow

File: src/scrapers/ibmCatalog.ts [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+      })
+    })
+  });
+  return products;
+}
+
+async function scrape(): Promise<void> {
+  const start = queryCatalogItems.bind(null, serviceParams, parsePricingJson);
+  const results = await download(start);
+  writeFile(filename, JSON.stringify(results, null, 2));
```
Comment (balazs-marjan) [inline src/scrapers/ibmCatalog.ts:505]: these promises are not awaited

File: src/scrapers/ibmCatalog.ts [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+  const products = parseProducts(results as CatalogJson[]);
+  writeFile('products.json', JSON.stringify(products, null, 2));
```
Comment (balazs-marjan) [inline src/scrapers/ibmCatalog.ts:507]: these promises are not awaited

**Review by balazs-marjan — CHANGES_REQUESTED** [review]
I adjusted the PR to use my apikey for production, I would get an error otherwise. I got a lot of ECONNRESETS, I'm fairly certain the script is hitting ratelimits :thinking:
Screenshot 2022-06-29 at 11 03 13

File: src/scrapers/ibmCatalog.ts [hunk 4]
```diff
@@ -0,0 +1,514 @@
+import GlobalCatalogV1, { CatalogEntry, PricingGet } from '@ibm-cloud/platform-services/global-catalog/v1';
+import { writeFile } from 'fs/promises';
+
+import { Product, Price } from '../db/types';
+import { generateProductHash } from '../db/helpers';
+import { upsertProducts } from '../db/upsert';
+import config from '../config';
+
+
+const client = GlobalCatalogV1.newInstance({});
```
[the remark this answers is not in the document]
Comment (balazs-marjan) [reply src/scrapers/ibmCatalog.ts:10]: It did for me

File: src/scrapers/ibmCatalog.ts [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+  const {
+    price: startPrice,
+    quantity_tier: startQuantityTier,
+  } = startingPrice;
+  
+  const attributes: ibmAttributes = {
+    planName,
+    planType: type,
+    startPrice: String(startPrice),
+    startQuantityTier: String(startQuantityTier),
```
Comment (balazs-marjan) [inline src/scrapers/ibmCatalog.ts:416]: I'm fairly certain this code tries to pull from the wrong place or just doesn't do anything
For every pasted product instance I get undefined values. i.e:
```
    "productHash": "b467373fa8e87482c57ccc27fcf68fbd",
    "sku": "datacatalog-professional",
    "vendorName": "ibm",
    "region": "eu-de",
    "service": "datacatalog",
    "productFamily": "service",
    "attributes": {
      "planName": "professional",
      "planType": "paygo",
      "startPrice": "undefined",
      "startQuantityTier": "undefined",
      "region": "eu-de"
    },
```

File: package-lock.json [hunk 6]
```diff
@@ -3228,6 +3418,56 @@
         "node": ">=6"
       }
     },
+    "node_modules/expect": {
+      "version": "26.6.2",
+      "resolved": "https://na.artifactory.swg-devops.com/artifactory/api/npm/wcp-tmp-ace-fr-team-npm-virtual/expect/-/expect-26.6.2.tgz?dl=https%3A%2F%2Fregistry.npmjs.org%2Fexpect%2F-%2Fexpect-26.6.2.tgz",
```
[the remark this answers is not in the document]
Comment (balazs-marjan) [reply package-lock.json:3423]: I think internal artifactory might still be a problem

**Review by balazs-marjan — APPROVED** [review]

**Comment by balazs-marjan** [comment]
Actually, how will we handle the conflicting records from the Kubernetes API and the GC entries?

**Review by sonkatomi — APPROVED** [review]
